src/EventStore.Core/Caching/DynamicCacheManager.cs

```diff
@@ -0,0 +1,200 @@
+
+			if (keepFreeMemBytes < 0)
+				throw new ArgumentException($"{nameof(keepFreeMemBytes)} must be non-negative.");
+
+			var dynamicWeightsPositive = cachesSettings
+				.Where(x => x.IsDynamic)
+				.All(x => x.Weight > 0);
+
+			if (!dynamicWeightsPositive)
+				throw new ArgumentException("Weight of all dynamic caches should be positive.");
+
+			_bus = bus;
+			_getFreeMem = getFreeMem;
+			_totalMem = totalMem;
+			_keepFreeMemPercent = keepFreeMemPercent;
+			_keepFreeMemBytes = keepFreeMemBytes;
+			_monitoringInterval = monitoringInterval;
+			_minResizeInterval = minResizeInterval;
+			_cachesSettings = cachesSettings;
+			_totalWeight = cachesSettings
+				.Where(x => x.IsDynamic)
+				.Sum(x => x.Weight);
+			_maxMemAllocation = new long[cachesSettings.Length];
+			Array.Fill(_maxMemAllocation, -1);
+
+			InitCacheSizes();
+			Tick();
+		}
+
+		public void Handle(MonitoringMessage.DynamicCacheManagerTick message) {
+			ThreadPool.QueueUserWorkItem(_ => {
+				try {
+					lock (_lock) { // only to add read/write barriers
+						ResizeCachesIfNeeded();
+					}
+				} finally {
+					Tick();
+				}
+			});
+		}
+
+		public void Handle(MonitoringMessage.InternalStatsRequest message) {
+			Thread.MemoryBarrier(); // just to ensure we're seeing latest values
+
+			var stats = new Dictionary<string, object>();
+
+			for (int i = 0; i < _cachesSettings.Length; i++) {
+				var statNamePrefix = $"es-cache-{_cachesSettings[i].Name}-";
+				stats[statNamePrefix + "name"] = _cachesSettings[i].Name;
+				stats[statNamePrefix + "weight"] = _cachesSettings[i].Weight;
+				stats[statNamePrefix + "mem-used"] = _cachesSettings[i].GetMemoryUsage();
+				stats[statNamePrefix + "mem-minAlloc"] = _cachesSettings[i].MinMemAllocation;
+				stats[statNamePrefix + "mem-maxAlloc"] = Interlocked.Read(ref _maxMemAllocation[i]);
+			}
+
+			message.Envelope.ReplyWith(new MonitoringMessage.InternalStatsRequestResponse(stats));
+		}
+
+		private void InitCacheSizes() {
+			var availableMem = CalcAvailableMemory(_getFreeMem(), 0L);
+			var cacheIndex = -1;
+			foreach (var cacheSettings in _cachesSettings) {
+				cacheIndex++;
+
+				if (!cacheSettings.IsDynamic) {
+					Log.Information("{name} cache size configured to ~{configuredMem:N0} bytes.",
+						cacheSettings.Name, cacheSettings.InitialMaxMemAllocation);
+					_maxMemAllocation[cacheIndex] = cacheSettings.InitialMaxMemAllocation;
+				} else {
+					var allocatedMem = CalcMemAllocation(availableMem, cacheSettings.Weight, cacheSettings.MinMemAllocation);
+					cacheSettings.InitialMaxMemAllocation = allocatedMem;
+					Log.Information(
+						"{name} cache size auto-configured to ~{allocatedMem:N0} bytes.",
+						cacheSettings.Name, allocatedMem);
+					_maxMemAllocation[cacheIndex] = allocatedMem;
+				}
+			}
+		}
+
+		private void ResizeCachesIfNeeded() {
+			var freeMem = _getFreeMem();
+			var keepFreeMem = Math.Max(_keepFreeMemBytes, _totalMem * _keepFreeMemPercent / 100);
```

Comment:
consider refactor with the same line in CalcAvailableMemory

Reply:
fixed in: c2d1aac9ea0cc79ff2f0ae338223c45a936a4e82